src/main/java/com/google/api/codegen/gapic/GapicGeneratorApp.java

```diff
@@ -99,10 +101,33 @@
 
   private ArtifactType artifactType;
 
+  private final boolean writeAsProtoResponse;
+
+  private CodeGeneratorResponse codeGeneratorProtoResponse = null;
+
+  public CodeGeneratorResponse getCodeGeneratorProtoResponse() {
+    return codeGeneratorProtoResponse;
+  }
+
   /** Constructs a code generator api based on given options. */
   public GapicGeneratorApp(ToolOptions options, ArtifactType artifactType) {
     super(options);
     this.artifactType = artifactType;
+    this.writeAsProtoResponse = false;
+  }
+
+  /**
+   * Constructs a code generator api based on given options.
+   *
+   * @param writeAsProtoResponse : If the output should be returned as a CodeGeneratorResponse,
+   *     instead of being written to file. The response object will be populated in
+   *     getCodeGeneratorProtoResponse().
+   */
+  public GapicGeneratorApp(
+      ToolOptions options, ArtifactType artifactType, boolean writeAsProtoResponse) {
```

Comment:
Rather than add a boolean flag and a codeGeneratorProtoResponse property to extract the data, what about abstracting out the writer concept. So accept a GapicWriter (or something like that) interface with a default implementation that writes to files, and an alternative impl that creates the proto response object. That way, the GapicGeneratorApp doesn't need to know anything about the proto response format, and writing to any new format (e.g. zip files, from the existing TODO) is easy in future. It could also be used for testing GapicGeneratorApp, if that's useful.

Reply:
Sure, see https://github.com/googleapis/gapic-generator/pull/2546. That should be reviewed first before going any farther in this review.

Reply:
Resolved.